commit 7e5a4b7e51d001d53df53f45833263bd1f26fe6f
Author: Thomas Edwards <tmfksoft@gmail.com>
Date:   Fri Jan 20 01:56:23 2017 +0000

    Removed some JS that interefered with tabs .active

diff --git a/src/js/ui.jsx b/src/js/ui.jsx
index ca75258..e4e8689 100644
--- a/src/js/ui.jsx
+++ b/src/js/ui.jsx
@@ -5,11 +5,6 @@ import * as data from "./data";
 export function initializeUI() {
 
 	$('.button').button();
-	$('#tab-bar').find('li').hover(function () {
-		$(this).addClass('ui-state-hover');
-	}, function () {
-		$(this).removeClass('ui-state-hover');
-	});
 
 	initializeCollapseControls();
 	initializeTimeSelector();
diff --git a/template/content.php b/template/content.php
index aacb2bf..324a360 100644
--- a/template/content.php
+++ b/template/content.php
@@ -20,15 +20,15 @@ if (!$section) $section = 'lagsummary';
 <div id="content">
 	<?php ad_banner_top_right(); ?>
 	<!-- http://refills.bourbon.io/#accordion-tabs -->
-	<div id="tab-bar" class="ui-tabs ui-widget ui-widget-content ui-corner-all">
-		<div class="tabs ui-tabs-nav ui-helper-reset ui-helper-clearfix ui-widget-header ui-corner-all">
+	<div id="tab-bar">
+		<div class="tabs">
 			<div class="tab-title <?=$section==="lag"?' active':''?>">
 				<a href="<?=util::buildurl(['section'=>'lag'])?>" class="tab ui-tabs-anchor">Lag Tree View</a>
 			</div>
 			<div class="tab-title <?=$section==="all"?' active':''?>">
 				<a href="<?=util::buildurl(['section'=>'all'])?>" class="tab ui-tabs-anchor">All Tree View</a>
 			</div>
-			<div class="tab-title <?=$section==="lagsummary"?' ui-tabs-active ui-state-active':''?>">
+			<div class="tab-title <?=$section==="lagsummary"?' active':''?>">
 				<a href="<?=util::buildurl(['section'=>'lagsummary'])?>" class="tab ui-tabs-anchor">Lag Summary View</a>
 			</div>
 			<div class="tab-title <?=$section==="summary"?' active':''?>">
@@ -40,7 +40,7 @@ if (!$section) $section = 'lagsummary';
 			<div class="tab-title <?=$section==="config"?' active':''?>">
 				<a href="<?=util::buildurl(['section'=>'config'])?>" class="tab ui-tabs-anchor">Config</a>
 			</div>
-			<div class="tab-title ui-state-default ui-corner-top <?=$section==="plugins"?' ui-tabs-active ui-state-active':''?>">
+			<div class="tab-title <?=$section==="plugins"?' ui-tabs-active ui-state-active':''?>">
 				<a href="<?=util::buildurl(['section'=>'plugins'])?>" class="tab ui-tabs-anchor">Plugin List</a>
 			</div>
 		</div>
